system/jlib/jdebug.cpp

```diff
@@ -1860,33 +1860,44 @@ void OsDiskInfo::initMajorMinor()
     Owned<IPipeProcess> pipe = createPipeProcess();
     if (pipe->run("list disks", cmd, nullptr, false, true, true, 8192))
     {
-        StringBuffer output;
-        Owned<ISimpleReadStream> pipeReader = pipe->getOutputStream();
-        readSimpleStream(output, *pipeReader);
-        unsigned exitcode = pipe->wait();
-        if ( (exitcode == 0) && (output.length() > 0) )
+        bool timedout = false;
+        // handle when lsblk runs too long
+        unsigned exitcode = pipe->wait(2000, timedout);
```

Comment:
You may need to read the output stream on another thread to avoid this I think.
Or even better would be to add that capability as an option to the IPipeProcess implementation (it already does a similar thing for stderr for similar reasons).

Reply:
@jakesmith I agree createPipeProcess() could be changed/improved to read stdout non-blocking (or at least check to see if there is data available before reading to prevent blocking) so it can read while also waiting, but I felt this was a different/larger issue.
I felt this issue was specifically to prevent the lsblk command called in the performance monitor to not run forever.
I do not see a new deadlock.  If the command outputs more than the pipe buffer can hold, additional writes will block - but that should not affect the wait() waiting on the (forkthread) thread to join.

Reply:
>additional writes will block - but that should not affect the wait()

if additional writes block, then the child process will be blocked, and the forkthread->join will be blocked because forkthread will be blocked in dowaitpid afaics.

Have you tried testing a child process that writes out a lot but should finish in less than the timeout?
Perhaps increase the timeout to something high for testing purposes and write a few 100K and see if it stalls (I think the pipe buffer  limit is 65K).

Reply:
I did test with a cmd that spews bytes to stdout as fast as possible.  But I will re-test and make sure it writes out more than the pipe buffer to verify.

Reply:
I have verified it does not block if cmd outputs more than pipe buffer size, it ends after the timeout.
But as mentioned above it does not get the output properly if 1) the cmd runs too long or 2) the cmd outputs more than the pipe buffer size.
So I will close this PR and refactor either the parent routine or the createPipeProcess() to get output while waiting.

Reply:
>it ends after the timeout.

yes, that is what I expected, didn't mean to imply deadlocked for ever, only until timeout, i.e. program could have finished v. quickly, but blocked until whatever the wait timeout is because of the write block on pipe buffer full with nothing reading it.